ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs

```diff
@@ -263,36 +322,73 @@ txSubmissionInbound _tracer maxUnacked TxSubmissionMempoolWriter{..} =
     serverReqTxs :: forall (n :: N).
                     Nat n
                  -> ServerState txid tx
-                 -> ServerStIdle n txid tx m ()
-    serverReqTxs n st =
-        SendMsgRequestTxsPipelined
+                 -> m (ServerStIdle n txid tx m ())
+    serverReqTxs n st = do
+        -- We can pick txs out of order if we want to, and we can decide not
+        -- to pick some txs at all (if for example we have seen them already).
+        -- We don't have to pick them all in one go, we can stick to a limit
+        -- like maxTxToRequest.
+
+        -- TODO @intricate: Should this whole thing be an STM tx?
+        recentTxIds <- atomically $ readTVar recentTxIdsVar
+
+            -- Potential transactions to request now and ones to keep to
+            -- consider later.
+        let (potentialTxsToRequest, availableTxids') =
+              Map.splitAt (fromIntegral maxTxToRequest) (availableTxids st)
+
+            -- Transactions to not request at all because we've already
+            -- recently added them to the mempool.
+            txsToIgnore = Set.fromList $
+              RecentTxIds.intersection (Map.keys potentialTxsToRequest) recentTxIds
+
+            -- Transactions to request now.
+            txsToRequest = Map.withoutKeys potentialTxsToRequest txsToIgnore
```

Comment:
So `potentialTxsToRequest` are the first `maxTxToRequest` in `availableTxids` (ordered by `txid`, which has no real meaning). Then you filter out the ones already in `recentTxIds`. What if all of the `potentialTxsToRequest` are in `recentTxIds`? Then you wont request any txs, even though `availableTxids` might contain txs not in `recentTxIds`. You need something like `take maxTxToRequest $ filter <not in recentTxIds> (availableTxids st)`.

(I'll continue with the rest of this PR tomorrow.)

Reply:
Ah, right.

Reply:
Looks like you resolved this